### PR TITLE
ACS-2503: Including Azure Blob in DAU design.

### PR DESCRIPTION
**Pull request by mpichura**
This PR updates the design of direct access URL feature to include Azure Blob.
The design is subject to small (hopefully) changes resulting from coming implementation.

### CODE REVIEW

File: docs/direct-access-urls.md [hunk 1]
```diff
@@ -8,16 +8,21 @@ ___
...
-for sharing objects. This feature is a perfect candidate for the implementation of direct
+for sharing objects.
+Azure Blob Storage provide [Shared Access Signature (SAS) tokens](https://docs.microsoft.com/en-us/rest/api/storageservices/delegate-access-with-shared-access-signature) which as a part of object's URL can serve very similar role.
+
+Above described features are perfect candidates for the implementation of direct
 access to our content (e.g. shortcutting the Shared File Store for transformations or
 faster/direct access in the context of ADW to documents).
 
->:warning: **Note:** The AWS S3 pre-signed URLs are temporary links with an expiration time.
+>:warning: **Note:** The AWS S3 pre-signed URLs are temporary links with an expiration time. Same counts for Azure SAS tokens which have expiration time.
```
Comment (montgolfiere) [inline docs/direct-access-urls.md:22]: Trivial: maybe replace "Same counts for" with "The same applies to" ?

File: docs/direct-access-urls.md [hunk 2]
```diff
@@ -8,16 +8,21 @@ ___
...
 
 The repository infrastructure now supports direct access urls. This includes the ContentService
```
Comment (montgolfiere) [inline docs/direct-access-urls.md:24]: Trivial: maybe add "Since ACS 7.1.0 the repository infrastructure ..." ?

File: docs/direct-access-urls.md [hunk 3]
```diff
@@ -8,16 +8,21 @@ ___
...
-and the ContentStore interface for which default methods have been provided so that ContentStore implementations that 
+and the ContentStore interface for which default methods have been provided so that ContentStore implementations that
 implement the old version of this interface will throw a Not Supported exception.
 The new methods are auditable using the node reference and time in seconds for which the direct access URL is
 valid for as the  parameters.
```
Comment (montgolfiere) [inline docs/direct-access-urls.md:28]: Trivial: reword "is valid for as the parameters." slightly (and remove extra space) ?

File: docs/direct-access-urls.md [hunk 4]
```diff
@@ -96,14 +101,21 @@ _caching content store_), should have dedicated configuration options:
     - If not set, the default system-wide setting will be used
       (`system.directAccessUrl.maxExpiryTimeInSec`).
 
+>**Note:** `{$connectorName}` depends on Alfresco Cloud Connector. For S3 Connector it is 's3', for Azure Connector it is 'az'.
+
 >**Note:** Callers within the platform (i.e. Java interfaces) can either request a specific
 > expiry time or rely on the default.
```
Comment (montgolfiere) [inline docs/direct-access-urls.md:107]: Minor: Do we clear state the OOTB (out-of-the-box) default values ? 

For example, what is the setting for:

- system.directAccessUrl.defaultExpiryTimeInSec ?
- connector.{$connectorName}.directAccessUrl.defaultExpiryTimeInSec (is it 30 secs as above for both s3 and az) ?


Either add here (or link to config code) ?

File: docs/direct-access-urls.md [hunk 5]
```diff
@@ -96,14 +101,21 @@ _caching content store_), should have dedicated configuration options:
...
 > be configured with either the default (common) S3 Connector-specific properties (i.e.
-> `connector.s3.directAccessUrl.enabled` & Co) OR new separate properties could be defined for
-> each and every store (e.g. `connector.s3store1.directAccessUrl.enabled`,
-> `connector.s3store2.directAccessUrl.enabled`, etc.).
+> `connector.s3.directAccessUrl.enabled` & common/default settings) OR new separate properties could be defined for
+> each and every store (e.g. `connector.s3.store1.directAccessUrl.enabled`,
+> `connector.s3.store2.directAccessUrl.enabled`, etc.).
+>
+> For multiple Azure blob containers, each of them can also be configured similarly to S3 buckets (i.e.
+> `connector.azure.directAccessUrl.enabled` & common/default settings) OR new separate properties could be defined for
```
Comment (montgolfiere) [inline docs/direct-access-urls.md:116]: Is "connector.azure.directAccessUrl.enabled" correct or should it be "connector.az.directAccessUrl.enabled" ?

File: docs/direct-access-urls.md [hunk 6]
```diff
@@ -142,16 +154,16 @@ The following endpoints can be used to obtain direct access URLs:
 * `/nodes/{nodeId}/versions/{versionId}/request-direct-access-url`
```
Comment (montgolfiere) [inline docs/direct-access-urls.md:154]: Is this complete ? eg.

`POST /nodes/{nodeId}/request-direct-access-url`
`POST /nodes/{nodeId}/renditions/{renditionId}/request-direct-access-url`

Maybe also prefix others with POST ?

Also @SaraAspery may also separately update again with new (if implemented):

`POST /nodes/{nodeId}/versions/{versionId}/renditions/{renditionId}/request-direct-access-url`

File: docs/direct-access-urls.md [hunk 7]
```diff
@@ -142,16 +154,16 @@ The following endpoints can be used to obtain direct access URLs:
...
 * `/deleted-nodes/{nodeId}/request-direct-access-url`
 * `/deleted-nodes/{nodeId}/renditions/{renditionId}/request-direct-access-url`
-  
+
 **Method:** **`POST`**
-  
+
 **Response:** Link to the resource wrapped in a JSON Object which also contains an attachment flag and the DAU expiration date.
```
Comment (montgolfiere) [inline docs/direct-access-urls.md:160]: Give example for S3 vs AZ ?

File: docs/direct-access-urls.md [hunk 8]
```diff
@@ -142,16 +154,16 @@ The following endpoints can be used to obtain direct access URLs:
...
-  
+
 **Error Codes:**
-* If there’s no Direct Access URL provider (e.g. Alfresco S3 Connector extension) installed in
+* If there’s no Direct Access URL provider (e.g. Alfresco S3 Connector / Azure Connector extension) installed in
```
Comment (montgolfiere) [inline docs/direct-access-urls.md:163]: Trivial: s/extension/module ?

File: docs/direct-access-urls.md [hunk 9]
```diff
@@ -182,6 +194,17 @@ The pre-signed request generates a download for the remote content.
 >**Known Limitations:** DAU generation on S3 depends on the security credentials used:
 > https://docs.aws.amazon.com/AmazonS3/latest/dev/ShareObjectPreSignedURL.html
 
+
+##### Azure Connector
```
Comment (montgolfiere) [inline docs/direct-access-urls.md:198]: This looks useful. Also add similar section for S3 Connector (with links to S3 docs) ?

File: docs/direct-access-urls.md [hunk 10]
```diff
@@ -231,19 +254,19 @@ ___
 
 ## Performance and scalability
 Each call to one of the new `.../request-direct-access-url` Alfresco REST endpoints results in the
-creation of a new (and separate) pre-signed URL for an AWS S3 object.
+creation of a new (and separate) pre-signed URL for an AWS S3 / Azure Blob object.
 
-The generation of a pre-signed URL is a purely AWS SDK **client-side operation** - meaning the
+The generation of a pre-signed URL is a purely AWS/Azure (TBC) SDK **client-side operation** - meaning the
```
Comment (montgolfiere) [inline docs/direct-access-urls.md:259]: Do we need TBC (we'll need to remember to come back and remove it) ? Can we check via experimental impl ?

**Review by montgolfiere — COMMENTED** [review]
LGTM. Mostly trivial / minor comments (please feel free to ignore any that are too trivial).

File: docs/direct-access-urls.md [hunk 11]
```diff
@@ -8,16 +8,21 @@ ___
...
 ## Overview
+Most Cloud Storage Providers allow generating publicly accessible URLs for sharing access to objects although the way it is provided and named differs throughout platforms.
+
 AWS S3 provides a way of generating
 [pre-signed URLs](https://docs.aws.amazon.com/AmazonS3/latest/dev/ShareObjectPreSignedURL.html)
-for sharing objects. This feature is a perfect candidate for the implementation of direct
+for sharing objects.
+Azure Blob Storage provide [Shared Access Signature (SAS) tokens](https://docs.microsoft.com/en-us/rest/api/storageservices/delegate-access-with-shared-access-signature) which as a part of object's URL can serve very similar role.
+
+Above described features are perfect candidates for the implementation of direct
```
[the remark this answers is not in the document]
Comment (mpichura) [reply docs/direct-access-urls.md:18]: Thanks, change is incoming.

File: docs/direct-access-urls.md (hunk 1 above)
Comment (mpichura) [reply docs/direct-access-urls.md:22]: Thanks, change is incoming.

File: docs/direct-access-urls.md (hunk 2 above)
Comment (mpichura) [reply docs/direct-access-urls.md:24]: Thanks, change is incoming.

File: docs/direct-access-urls.md (hunk 3 above)
Comment (mpichura) [reply docs/direct-access-urls.md:28]: Rephrased part is coming in the next commit.

File: docs/direct-access-urls.md (hunk 4 above)
Comment (mpichura) [reply docs/direct-access-urls.md:107]: In the sections above there are proposed default values for all of these properties (30 seconds). I added a reminder about that here.
I think that adding links to config code may make some mess in the future when the code gets changed

File: docs/direct-access-urls.md (hunk 5 above)
Comment (mpichura) [reply docs/direct-access-urls.md:116]: Thanks for catching that - change incoming.

File: docs/direct-access-urls.md (hunk 6 above)
Comment (mpichura) [reply docs/direct-access-urls.md:154]: I recon that only versions-renditions was missing which I added. 
About adding POST prefix to all - I guess that just in the next line after `Paths` there is a paragraph explaining that all of them are POST methods:
`**Method:** **`POST`**`

File: docs/direct-access-urls.md (hunk 7 above)
Comment (mpichura) [reply docs/direct-access-urls.md:160]: Examples added.

File: docs/direct-access-urls.md (hunk 8 above)
Comment (mpichura) [reply docs/direct-access-urls.md:163]: Reworded to module - hope that's what you meant.

File: docs/direct-access-urls.md (hunk 10 above)
Comment (mpichura) [reply docs/direct-access-urls.md:259]: Checked - no TBC needed.

File: docs/direct-access-urls.md (hunk 9 above)
Comment (mpichura) [reply docs/direct-access-urls.md:198]: Looks like similar section is right above or I misunderstood you.

File: docs/direct-access-urls.md [hunk 12]
```diff
@@ -140,18 +150,39 @@ The following endpoints can be used to obtain direct access URLs:
 * `/nodes/{nodeId}/request-direct-access-url`
 * `/nodes/{nodeId}/renditions/{renditionId}/request-direct-access-url`
 * `/nodes/{nodeId}/versions/{versionId}/request-direct-access-url`
+* `/nodes/{nodeId}/versions/{versionId}/renditions/{renditionId}/request-direct-access-url`
```
[the remark this answers is not in the document]
Comment (mpichura) [reply docs/direct-access-urls.md]: Removed.

File: docs/direct-access-urls.md (hunk 6 above)
Comment (SaraAspery) [reply docs/direct-access-urls.md:152]: Thanks for that :)

File: docs/direct-access-urls.md (hunk 4 above)
Comment (montgolfiere) [reply docs/direct-access-urls.md]: Thanks. 

I might have missed it but I couldn't see the default (OOTB) value for 'system.directAccessUrl.maxExpiryTimeInSec'.

File: docs/direct-access-urls.md (hunk 7 above)
Comment (montgolfiere) [reply docs/direct-access-urls.md:158]: Great, thanks 👍

File: docs/direct-access-urls.md (hunk 9 above)
Comment (montgolfiere) [reply docs/direct-access-urls.md:216]: Yes, you're right - I missed it.

**Review by montgolfiere — APPROVED** [review]

**Review by SaraAspery — APPROVED** [review]
Looks great :)

**Pull request merged**